scipy/sparse/csgraph/_laplacian.py

```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+           [-3,  8, -5],
+           [-6, -7, 13]])
+    >>> L_in_degree + L_out_degree.T
+    array([[ 12,  -4,  -8],
+            [ -4,  16, -12],
+            [ -8, -12,  20]])
+    >>> csgraph.laplacian(G, symmetrized=True)
+    array([[ 12,  -4,  -8],
+           [ -4,  16, -12],
+           [ -8, -12,  20]])
+    >>> csgraph.laplacian(G + G.T)
+    array([[ 12,  -4,  -8],
+           [ -4,  16, -12],
+           [ -8, -12,  20]])
     """
     if csgraph.ndim != 2 or csgraph.shape[0] != csgraph.shape[1]:
-        raise ValueError('csgraph must be a square matrix or array')
+        raise ValueError("csgraph must be a square matrix or array")
 
-    if normed and (np.issubdtype(csgraph.dtype, np.signedinteger)
-                   or np.issubdtype(csgraph.dtype, np.uint)):
-        csgraph = csgraph.astype(float)
+    if normed and (
+        np.issubdtype(csgraph.dtype, np.signedinteger)
+        or np.issubdtype(csgraph.dtype, np.uint)
+    ):
+        csgraph = csgraph.astype(np.float64)
 
     create_lap = _laplacian_sparse if isspmatrix(csgraph) else _laplacian_dense
     degree_axis = 1 if use_out_degree else 0
-    lap, d = create_lap(csgraph, normed=normed, axis=degree_axis,
-                        copy=copy)
+
+    lap, d = create_lap(
+        csgraph,
+        normed=normed,
+        axis=degree_axis,
+        copy=copy,
+        form=form,
+        dtype=dtype,
+        symmetrized=symmetrized,
+    )
     if return_diag:
         return lap, d
     return lap
 
 
-def _setdiag_dense(A, d):
-    A.flat[::len(d)+1] = d
+def _setdiag_dense(m, d):
+    m.flat[:: len(d) + 1] = d
 
 
-def _laplacian_sparse(graph, normed=False, axis=0,
-                      copy=True):
-    needs_copy = False
-    if graph.format in ('lil', 'dok'):
-        m = graph.tocoo()
-    else:
-        m = graph
-        if copy:
-            needs_copy = True
-    w = m.sum(axis=axis).getA1() - m.diagonal()
-    if normed:
-        m = m.tocoo(copy=needs_copy)
-        isolated_node_mask = (w == 0)
-        w = np.where(isolated_node_mask, 1, np.sqrt(w))
-        m.data /= w[m.row]
-        m.data /= w[m.col]
-        m.data *= -1
-        m.setdiag(1 - isolated_node_mask)
+def _md(m, d):
+    return lambda v: v * d[np.newaxis, :] - m @ v
+
+
+def _md_normed(m, d, nd):
+    md_sym = _md(m, d)
+    return lambda v: nd * md_sym(v) * nd[:, np.newaxis]
+
+
+def _md_sym(m, d):
+    return (
+        lambda v: v * d[np.newaxis, :]
+        - m @ v
+        - np.transpose(np.conjugate(np.transpose(np.conjugate(v)) @ m))
+    )
+
+
+def _md_normed_sym(m, d, nd):
+    md_sym = _md_sym(m, d)
+    return lambda v: nd * md_sym(v) * nd[:, np.newaxis]
+
+
+def _linearoperator(mv, shape, dtype):
+    mv_as_lo = LinearOperator(matvec=mv, matmat=mv, shape=shape, dtype=dtype)
+    return mv_as_lo
+
+
+def _laplacian_sparse(graph, normed, axis, copy, form, dtype, symmetrized):
+    if dtype is None:
+        dtype = graph.dtype
+
+    if form != "array":
```

Comment:
As these are effectively different functions, I think it would be easier to read if you split these into _laplacian_sparse_matrix() for the form='array' case and _laplacian_sparse_function() for the other cases. It looks like the only duplicated code would be the default value for `dtype`.

Reply:
Done. I kept the names _laplacian_sparse and _laplacian_dense for the original functions for the array output and introduced _laplacian_sparse_flo and _laplacian_dense_flo. Please have a look.

There are now some unused variables in function calls, which I need to figure out how to handle